File: app/src/main/java/info/nightscout/androidaps/plugins/constraints/signatureVerifier/SignatureVerifierPlugin.java
package info.nightscout.androidaps.plugins.constraints.signatureVerifier;

import android.content.pm.PackageManager;
import android.content.pm.Signature;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.spongycastle.util.encoders.Hex;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URL;
import java.net.URLConnection;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.interfaces.Constraint;
import info.nightscout.androidaps.interfaces.ConstraintsInterface;
import info.nightscout.androidaps.interfaces.PluginBase;
import info.nightscout.androidaps.interfaces.PluginDescription;
import info.nightscout.androidaps.interfaces.PluginType;
import info.nightscout.androidaps.logging.L;
import info.nightscout.androidaps.plugins.bus.RxBus;
import info.nightscout.androidaps.plugins.general.overview.events.EventNewNotification;
import info.nightscout.androidaps.plugins.general.overview.notifications.Notification;
import info.nightscout.androidaps.utils.SP;

/**
 * AndroidAPS is meant to be build by the user.
 * In case someone decides to leak a ready-to-use APK nonetheless, we can still disable it.
 * Self-compiled APKs with privately held certificates cannot and will not be disabled.
 */
public class SignatureVerifierPlugin extends PluginBase implements ConstraintsInterface {

    private static final String REVOKED_CERTS_URL = "https://raw.githubusercontent.com/MilosKozak/AndroidAPS/master/app/src/main/assets/revoked_certs.txt";
    private static final long UPDATE_INTERVAL = TimeUnit.DAYS.toMillis(1);

    private static SignatureVerifierPlugin plugin = new SignatureVerifierPlugin();

    private Logger log = LoggerFactory.getLogger(L.CORE);
    private final Object $lock = new Object[0];
    private File revokedCertsFile;
    private List<byte[]> revokedCerts;

    public static SignatureVerifierPlugin getPlugin() {
        return plugin;
    }

    private SignatureVerifierPlugin() {
        super(new PluginDescription()
                .mainType(PluginType.CONSTRAINTS)
                .neverVisible(true)
                .alwaysEnabled(true)
                .showInList(false)
                .pluginName(R.string.signature_verifier));
    }

    @Override
    protected void onStart() {
        super.onStart();
        revokedCertsFile = new File(MainApp.instance().getFilesDir(), "revoked_certs.txt");
        new Thread(() -> {
            loadLocalRevokedCerts();
            if (shouldDownloadCerts()) {
                try {
                    downloadAndSaveRevokedCerts();
                } catch (IOException e) {
                    log.error("Could not download revoked certs", e);
                }
            }
            if (hasIllegalSignature()) showNotification();
        }).start();
    }

    @Override
    public Constraint<Boolean> isLoopInvocationAllowed(Constraint<Boolean> value) {
        if (hasIllegalSignature()) {
            showNotification();
            value.set(false);
        }
        if (shouldDownloadCerts()) {
            new Thread(() -> {
                try {
                    downloadAndSaveRevokedCerts();
                } catch (IOException e) {
                    log.error("Could not download revoked certs", e);
                }
            }).start();
        }
        return value;
    }

    private void showNotification() {
        Notification notification = new Notification(Notification.INVALID_VERSION, MainApp.gs(R.string.running_invalid_version), Notification.URGENT);
        RxBus.INSTANCE.send(new EventNewNotification(notification));
    }

    private boolean hasIllegalSignature() {
        try {
            synchronized ($lock) {
                if (revokedCerts == null) return false;
                Signature[] signatures = MainApp.instance().getPackageManager().getPackageInfo(MainApp.instance().getPackageName(), PackageManager.GET_SIGNATURES).signatures;
                if (signatures != null) {
                    for (Signature signature : signatures) {
                        MessageDigest digest = MessageDigest.getInstance("SHA256");
                        byte[] fingerprint = digest.digest(signature.toByteArray());
                        for (byte[] cert : revokedCerts) {
                            if (Arrays.equals(cert, fingerprint)) {
                                return true;
                            }
                        }
                    }
                }
            }
        } catch (PackageManager.NameNotFoundException | NoSuchAlgorithmException e) {
            log.error("Error in SignatureVerifierPlugin", e);
        }
        return false;
    }

    public List<String> shortHashes() {
        List<String> hashes = new ArrayList<>();
        try {
            Signature[] signatures = MainApp.instance().getPackageManager().getPackageInfo(MainApp.instance().getPackageName(), PackageManager.GET_SIGNATURES).signatures;
            if (signatures != null) {
                for (Signature signature : signatures) {
                    MessageDigest digest = MessageDigest.getInstance("SHA256");
                    byte[] fingerprint = digest.digest(signature.toByteArray());
                    String hash = Hex.toHexString(fingerprint);
                    log.debug("Found signature: " + hash);
                    log.debug("Found signature (short): " + singleCharMap(fingerprint));
                    hashes.add(singleCharMap(fingerprint));
                }
            }
        } catch (PackageManager.NameNotFoundException | NoSuchAlgorithmException e) {
            log.error("Error in SignatureVerifierPlugin", e);
        }
        return hashes;
    }

    String map = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!\"§$%&/()=?,.-;:_<>|°^`´\\@€*'#+~{}[]¿¡áéíóúàèìòùöäü`ÁÉÍÓÚÀÈÌÒÙÖÄÜßÆÇÊËÎÏÔŒÛŸæçêëîïôœûÿĆČĐŠŽćđšžñΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡ\u03A2ΣΤΥΦΧΨΩαβγδεζηθικλμνξοπρςστυφχψωϨϩϪϫϬϭϮϯϰϱϲϳϴϵ϶ϷϸϹϺϻϼϽϾϿЀЁЂЃЄЅІЇЈЉЊЋЌЍЎЏАБВГДЕЖЗ";

    private String singleCharMap(byte[] array) {
        StringBuilder sb = new StringBuilder();
        for (byte b : array) {
            sb.append(map.charAt(b & 0xFF));
        }
        return sb.toString();
    }

    public String singleCharUnMap(String shortHash) {
        byte[] array = new byte[shortHash.length()];
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < array.length; i++) {
            if (i != 0) sb.append(":");
            sb.append(String.format("%02X", 0xFF & map.charAt(map.indexOf(shortHash.charAt(i)))));
        }
        return sb.toString();
    }

    private boolean shouldDownloadCerts() {
        return System.currentTimeMillis() - SP.getLong(R.string.key_last_revoked_certs_check, 0L) >= UPDATE_INTERVAL;
    }

    private void downloadAndSaveRevokedCerts() throws IOException {
        String download = downloadRevokedCerts();
        saveRevokedCerts(download);
        SP.putLong(R.string.key_last_revoked_certs_check, System.currentTimeMillis());
        synchronized ($lock) {
            revokedCerts = parseRevokedCertsFile(download);
        }
    }

    private void loadLocalRevokedCerts() {
        try {
            String revokedCerts = readCachedDownloadedRevokedCerts();
            if (revokedCerts == null) revokedCerts = readRevokedCertsInAssets();
            synchronized ($lock) {
                this.revokedCerts = parseRevokedCertsFile(revokedCerts);
            }
        } catch (IOException e) {
            log.error("Error in SignatureVerifierPlugin", e);
        }
    }

    private void saveRevokedCerts(String revokedCerts) throws IOException {
        OutputStream outputStream = new FileOutputStream(revokedCertsFile);
        outputStream.write(revokedCerts.getBytes(StandardCharsets.UTF_8));
        outputStream.close();
    }

    private String downloadRevokedCerts() throws IOException {
        URLConnection connection = new URL(REVOKED_CERTS_URL).openConnection();
        return readInputStream(connection.getInputStream());
    }

    private String readInputStream(InputStream inputStream) throws IOException {
        try {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            byte[] buffer = new byte[1024];
            int read;
            while ((read = inputStream.read(buffer)) != -1) {
                baos.write(buffer, 0, read);
            }
            baos.flush();
            return new String(baos.toByteArray(), StandardCharsets.UTF_8);
        } finally {
            inputStream.close();
        }
    }

    private String readRevokedCertsInAssets() throws IOException {
        InputStream inputStream = MainApp.instance().getAssets().open("revoked_certs.txt");
        return readInputStream(inputStream);
    }

    private String readCachedDownloadedRevokedCerts() throws IOException {
        if (!revokedCertsFile.exists()) return null;
        return readInputStream(new FileInputStream(revokedCertsFile));
    }

    private List<byte[]> parseRevokedCertsFile(String file) {
        List<byte[]> revokedCerts = new ArrayList<>();
        for (String line : file.split("\n")) {
            if (line.startsWith("#")) continue;
            revokedCerts.add(Hex.decode(line.replace(" ", "").replace(":", "")));
        }
        return revokedCerts;
    }
}

File: plugins/constraints/src/main/kotlin/app/aaps/plugins/constraints/signatureVerifier/SignatureVerifierPlugin.kt
package app.aaps.plugins.constraints.signatureVerifier

import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.HandlerThread
import app.aaps.core.data.plugin.PluginType
import app.aaps.core.interfaces.constraints.Constraint
import app.aaps.core.interfaces.constraints.PluginConstraints
import app.aaps.core.interfaces.logging.AAPSLogger
import app.aaps.core.interfaces.notifications.Notification
import app.aaps.core.interfaces.plugin.PluginBaseWithPreferences
import app.aaps.core.interfaces.plugin.PluginDescription
import app.aaps.core.interfaces.resources.ResourceHelper
import app.aaps.core.interfaces.ui.UiInteraction
import app.aaps.core.keys.interfaces.Preferences
import app.aaps.plugins.constraints.R
import app.aaps.plugins.constraints.signatureVerifier.keys.SignatureVerifierLongKey
import org.spongycastle.util.encoders.Hex
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.URL
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * AndroidAPS is meant to be build by the user.
 * In case someone decides to leak a ready-to-use APK nonetheless, we can still disable it.
 * Self-compiled APKs with privately held certificates cannot and will not be disabled.
 */
@Suppress("PrivatePropertyName")
@Singleton
class SignatureVerifierPlugin @Inject constructor(
    aapsLogger: AAPSLogger,
    rh: ResourceHelper,
    preferences: Preferences,
    private val context: Context,
    private val uiInteraction: UiInteraction
) : PluginBaseWithPreferences(
    pluginDescription = PluginDescription()
        .mainType(PluginType.CONSTRAINTS)
        .neverVisible(true)
        .alwaysEnabled(true)
        .showInList { false }
        .pluginName(R.string.signature_verifier),
    ownPreferences = listOf(SignatureVerifierLongKey::class.java),
    aapsLogger, rh, preferences
), PluginConstraints {

    private var handler: Handler? = null

    private val REVOKED_CERTS_URL = "https://raw.githubusercontent.com/nightscout/AndroidAPS/master/app/src/main/assets/revoked_certs.txt"
    private val UPDATE_INTERVAL = TimeUnit.DAYS.toMillis(1)

    private val lock: Any = arrayOfNulls<Any>(0)
    private var revokedCertsFile: File? = null
    private var revokedCerts: List<ByteArray>? = null
    override fun onStart() {
        super.onStart()
        handler = Handler(HandlerThread(this::class.simpleName + "Handler").also { it.start() }.looper)
        revokedCertsFile = File(context.filesDir, "revoked_certs.txt")
        handler?.post {
            loadLocalRevokedCerts()
            if (shouldDownloadCerts()) {
                try {
                    downloadAndSaveRevokedCerts()
                } catch (e: IOException) {
                    aapsLogger.error("Could not download revoked certs", e)
                }
            }
            if (hasIllegalSignature()) showNotification()
        }
    }

    override fun onStop() {
        handler?.removeCallbacksAndMessages(null)
        handler?.looper?.quit()
        handler = null
        super.onStop()
    }

    override fun isLoopInvocationAllowed(value: Constraint<Boolean>): Constraint<Boolean> {
        if (hasIllegalSignature()) {
            showNotification()
            value.set(false)
        }
        if (shouldDownloadCerts()) {
            handler?.post {
                try {
                    downloadAndSaveRevokedCerts()
                } catch (e: IOException) {
                    aapsLogger.error("Could not download revoked certs", e)
                }
            }
        }
        return value
    }

    private fun showNotification() {
        uiInteraction.addNotification(Notification.INVALID_VERSION, rh.gs(R.string.running_invalid_version), Notification.URGENT)
    }

    private fun hasIllegalSignature(): Boolean {
        try {
            synchronized(lock) {
                if (revokedCerts == null) return false
                // TODO Change after raising min API to 28
                @Suppress("DEPRECATION", "PackageManagerGetSignatures")
                val signatures = context.packageManager.getPackageInfo(context.packageName, PackageManager.GET_SIGNATURES).signatures
                if (signatures != null) {
                    for (signature in signatures) {
                        val digest = MessageDigest.getInstance("SHA256")
                        val fingerprint = digest.digest(signature.toByteArray())
                        for (cert in revokedCerts!!) {
                            if (cert.contentEquals(fingerprint)) {
                                return true
                            }
                        }
                    }
                }
            }
        } catch (e: PackageManager.NameNotFoundException) {
            aapsLogger.error("Error in SignatureVerifierPlugin", e)
        } catch (e: NoSuchAlgorithmException) {
            aapsLogger.error("Error in SignatureVerifierPlugin", e)
        }
        return false
    }

    fun shortHashes(): List<String> {
        val hashes: MutableList<String> = ArrayList()
        try {
            // TODO Change after raising min API to 28
            @Suppress("DEPRECATION", "PackageManagerGetSignatures")
            val signatures = context.packageManager.getPackageInfo(context.packageName, PackageManager.GET_SIGNATURES).signatures
            if (signatures != null) {
                for (signature in signatures) {
                    val digest = MessageDigest.getInstance("SHA256")
                    val fingerprint = digest.digest(signature.toByteArray())
                    val hash = Hex.toHexString(fingerprint)
                    aapsLogger.debug("Found signature: $hash")
                    aapsLogger.debug("Found signature (short): " + singleCharMap(fingerprint))
                    hashes.add(singleCharMap(fingerprint))
                }
            }
        } catch (e: PackageManager.NameNotFoundException) {
            aapsLogger.error("Error in SignatureVerifierPlugin", e)
        } catch (e: NoSuchAlgorithmException) {
            aapsLogger.error("Error in SignatureVerifierPlugin", e)
        }
        return hashes
    }

    @Suppress("SpellCheckingInspection")
    var map =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!\"§$%&/()=?,.-;:_<>|°^`´\\@€*'#+~{}[]¿¡áéíóúàèìòùöäü`ÁÉÍÓÚÀÈÌÒÙÖÄÜßÆÇÊËÎÏÔŒÛŸæçêëîïôœûÿĆČĐŠŽćđšžñΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡ\u03A2ΣΤΥΦΧΨΩαβγδεζηθικλμνξοπρςστυφχψωϨϩϪϫϬϭϮϯϰϱϲϳϴϵ϶ϷϸϹϺϻϼϽϾϿЀЁЂЃЄЅІЇЈЉЊЋЌЍЎЏАБВГДЕЖЗ"

    private fun singleCharMap(array: ByteArray): String {
        val sb = StringBuilder()
        for (b in array) {
            sb.append(map[b.toInt() and 0xFF])
        }
        return sb.toString()
    }

    private fun shouldDownloadCerts(): Boolean {
        return System.currentTimeMillis() - preferences.get(SignatureVerifierLongKey.LastRevokedCertCheck) >= UPDATE_INTERVAL
    }

    @Throws(IOException::class) private fun downloadAndSaveRevokedCerts() {
        val download = downloadRevokedCerts()
        saveRevokedCerts(download)
        preferences.put(SignatureVerifierLongKey.LastRevokedCertCheck, System.currentTimeMillis())
        synchronized(lock) { revokedCerts = parseRevokedCertsFile(download) }
    }

    private fun loadLocalRevokedCerts() {
        try {
            var revokedCerts = readCachedDownloadedRevokedCerts()
            if (revokedCerts == null) revokedCerts = readRevokedCertsInAssets()
            synchronized(lock) { this.revokedCerts = parseRevokedCertsFile(revokedCerts) }
        } catch (e: IOException) {
            aapsLogger.error("Error in SignatureVerifierPlugin", e)
        }
    }

    @Throws(IOException::class)
    private fun saveRevokedCerts(revokedCerts: String) {
        val outputStream: OutputStream = FileOutputStream(revokedCertsFile)
        outputStream.write(revokedCerts.toByteArray(StandardCharsets.UTF_8))
        outputStream.close()
    }

    @Throws(IOException::class) private fun downloadRevokedCerts(): String {
        val connection = URL(REVOKED_CERTS_URL).openConnection()
        return readInputStream(connection.getInputStream())
    }

    @Throws(IOException::class)
    fun readInputStream(inputStream: InputStream): String {
        return try {
            val os = ByteArrayOutputStream()
            val buffer = ByteArray(1024)
            var read: Int
            while (inputStream.read(buffer).also { read = it } != -1) {
                os.write(buffer, 0, read)
            }
            os.flush()
            String(os.toByteArray(), StandardCharsets.UTF_8)
        } finally {
            inputStream.close()
        }
    }

    @Throws(IOException::class) private fun readRevokedCertsInAssets(): String {
        val inputStream = context.assets.open("revoked_certs.txt")
        return readInputStream(inputStream)
    }

    @Throws(IOException::class)
    private fun readCachedDownloadedRevokedCerts(): String? {
        return if (!revokedCertsFile!!.exists()) null else readInputStream(FileInputStream(revokedCertsFile))
    }

    private fun parseRevokedCertsFile(file: String?): List<ByteArray> {
        val revokedCerts: MutableList<ByteArray> = ArrayList()
        for (line in file!!.split("\n").toTypedArray()) {
            if (line.startsWith("#")) continue
            revokedCerts.add(Hex.decode(line.replace(" ", "").replace(":", "")))
        }
        return revokedCerts
    }
}